pkg/approver/aws/approver.go

```diff
@@ -0,0 +1,147 @@
+package aws
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/juju/ratelimit"
+
+	certificates "k8s.io/api/certificates/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	certificatesv1b1listers "k8s.io/client-go/listers/certificates/v1beta1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+)
+
+const (
+	masterNodeRoleLabelKey = "node-role.kubernetes.io/master"
+	workerNodeRoleLabelKey = "node-role.kubernetes.io/node"
+)
+
+// Config defines settings for Approver.
+type Config struct {
+	RegionName  string   `json:"region_name"`
+	AllowedASGs []string `json:"allowed_auto_scaling_groups"`
+}
+
+// Approver approvers CSRs.
+type Approver struct {
+	kubeClient  kubernetes.Interface
+	aws         *awsCloud
+	allowedASGs sets.String
+
+	csrLister certificatesv1b1listers.CertificateSigningRequestLister
+	csrSynced cache.InformerSynced
+	queue     workqueue.RateLimitingInterface
+}
+
+// New return a new Approver.
+func New(client kubernetes.Interface, config Config) (*Approver, error) {
+	ar := &Approver{
+		kubeClient: client,
+	}
+
+	c, err := newAWSCloud(config.RegionName)
+	if err != nil {
+		return nil, err
+	}
+	ar.aws = c
+
+	ar.allowedASGs = sets.NewString(config.AllowedASGs...)
+
+	ar.queue = workqueue.NewRateLimitingQueue(workqueue.NewMaxOfRateLimiter(
+		workqueue.NewItemExponentialFailureRateLimiter(200*time.Millisecond, 100*time.Second),
+		&workqueue.BucketRateLimiter{Bucket: ratelimit.NewBucketWithRate(float64(10), int64(100))},
+	))
+
+	sharedInformer := informers.NewSharedInformerFactory(ar.kubeClient, 0)
+	csrInformer := sharedInformer.Certificates().V1beta1().CertificateSigningRequests()
+	csrInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			csr := obj.(*certificates.CertificateSigningRequest)
+			glog.V(4).Infof("Adding certificate request %s", csr.Name)
+			ar.enqueueCertificateRequest(obj)
+		},
+	})
+	ar.csrLister = csrInformer.Lister()
+	ar.csrSynced = csrInformer.Informer().HasSynced
+	go sharedInformer.Start(wait.NeverStop)
```

Comment:
nit: New should take a context so this can be stopped.

Reply:
upstream starts informers in main func before calling Run(). removed starting informers step in New.